overlord/snapstate/snapstate.go

```diff
@@ -782,10 +790,25 @@ func doInstall(st *state.State, snapst *SnapState, snapsup SnapSetup, compsups [
 	installSet := state.NewTaskSet(tasks...)
 	installSet.MarkEdge(prereq, BeginEdge)
 	installSet.MarkEdge(setupAliases, BeforeHooksEdge)
-	installSet.MarkEdge(setupSecurity, BeforeMaybeRebootEdge)
-	installSet.MarkEdge(linkSnap, MaybeRebootEdge)
-	installSet.MarkEdge(autoConnect, MaybeRebootWaitEdge)
-	installSet.MarkEdge(setAutoAliases, AfterMaybeRebootWaitEdge)
+
+	// restart related edges
+	installSet.MarkEdge(setupSecurity, BeforeMaybeRebootEdge) // this edge is just for tests
+	var rebootTask, waitRebootTask *state.Task
+	if setupKmodComponents == nil {
+		rebootTask = linkSnap
+		waitRebootTask = autoConnect
+	} else {
+		logger.Noticef("reboot will happen after set-up of kernel-modules")
+		rebootTask = setupKmodComponents
+		waitRebootTask = afterSetupKmodComps
+	}
+	installSet.MarkEdge(rebootTask, MaybeRebootEdge)
+	installSet.MarkEdge(waitRebootTask, MaybeRebootWaitEdge)
+	// Let tasks know if they have to do something about restarts
+	rebootTask.Set("prepare-restart", true)
+	waitRebootTask.Set("finish-restart", true)
+	installSet.MarkEdge(setAutoAliases, AfterMaybeRebootWaitEdge) // this edge is just for tests
```

Comment:
Same for this one, and just get the Halt task for `MaybeRebootWaitEdge`, but again, outside the scope of this PR.